File: Structural/Adapter/Adapter.java
// 目标接口
public interface Shape {
    void boundingBox();
    boolean isEmpty();
    void createManipulator();
}

// 已存在的、接口不兼容的类
public class TextView {
    public void getOrigin()   { System.out.println("TextView::GetOrigin"); }
    public void getExtent()   { System.out.println("TextView::GetExtent"); }
    public boolean isEmpty()  { return false; }
}

// 对象适配器
public class TextShape implements Shape {
    private final TextView textView;

    public TextShape(TextView tv) { this.textView = tv; }

    @Override
    public void boundingBox() {
        textView.getOrigin();
        textView.getExtent();
        System.out.println("Shape::BoundingBox (adapted from TextView)");
    }

    @Override
    public boolean isEmpty() {
        return textView.isEmpty();
    }

    @Override
    public void createManipulator() {
        System.out.println("Shape::CreateManipulator (TextManipulator)");
    }

    // Demo
    public static void main(String[] args) {
        Shape textShape = new TextShape(new TextView());
        textShape.boundingBox();
    }
}

File: Structural/Adapter/Adapter.py
from typing import Protocol

class Shape(Protocol):
    def bounding_box(self) -> None: ...
    def is_empty(self) -> bool: ...
    def create_manipulator(self) -> None: ...

class TextView:
    def get_origin(self) -> None:
        print("TextView::get_origin")
    def get_extent(self) -> None:
        print("TextView::get_extent")
    def is_empty(self) -> bool:
        return False

class TextShape:
    def __init__(self, text_view: TextView):
        self._tv = text_view

    def bounding_box(self) -> None:
        self._tv.get_origin()
        self._tv.get_extent()
        print("Shape::bounding_box (adapted from TextView)")

    def is_empty(self) -> bool:
        return self._tv.is_empty()

    def create_manipulator(self) -> None:
        print("Shape::create_manipulator (TextManipulator)")

if __name__ == "__main__":
    shape: Shape = TextShape(TextView())
    shape.bounding_box()